File: shared/src/commonTest/kotlin/com/ttcoachai/shared/drill/DrillMetricsTest.kt
package com.ttcoachai.shared.drill

import com.ttcoachai.shared.models.Coco17
import com.ttcoachai.shared.models.Handedness
import com.ttcoachai.shared.models.Keypoint2D
import com.ttcoachai.shared.models.PoseFrame2D
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class DrillMetricsTest {

    private fun frame(vararg overrides: Pair<Int, Keypoint2D>): PoseFrame2D {
        val kp = MutableList(17) { Keypoint2D(0.5f, 0.5f, 1f) }
        overrides.forEach { (i, p) -> kp[i] = p }
        return PoseFrame2D(0, 0L, kp)
    }

    @Test
    fun extractsAllFiveMetricsFromGoodFrame() {
        val f = frame(
            Coco17.RIGHT_SHOULDER to Keypoint2D(0.50f, 0.30f, 1f),
            Coco17.LEFT_SHOULDER to Keypoint2D(0.48f, 0.30f, 1f),
            Coco17.RIGHT_ELBOW to Keypoint2D(0.55f, 0.42f, 1f),
            Coco17.RIGHT_WRIST to Keypoint2D(0.65f, 0.40f, 1f),
            Coco17.RIGHT_HIP to Keypoint2D(0.50f, 0.55f, 1f),
            Coco17.LEFT_HIP to Keypoint2D(0.48f, 0.55f, 1f),
            Coco17.RIGHT_KNEE to Keypoint2D(0.50f, 0.72f, 1f),
            Coco17.RIGHT_ANKLE to Keypoint2D(0.52f, 0.90f, 1f)
        )
        val m = DrillMetrics.extractAtFrame(f, Handedness.RIGHT, 1f)
        assertEquals(
            setOf(
                DrillMetrics.METRIC_ELBOW_ANGLE,
                DrillMetrics.METRIC_SHOULDER_ANGLE,
                DrillMetrics.METRIC_KNEE_BEND,
                DrillMetrics.METRIC_TORSO_LEAN,
                DrillMetrics.METRIC_SHOULDER_TILT
            ),
            m.keys
        )
    }

    @Test
    fun lowScoreJointDropsOnlyAffectedMetrics() {
        val f = frame(
            Coco17.RIGHT_SHOULDER to Keypoint2D(0.50f, 0.30f, 1f),
            Coco17.LEFT_SHOULDER to Keypoint2D(0.48f, 0.30f, 1f),
            Coco17.RIGHT_ELBOW to Keypoint2D(0.55f, 0.42f, 1f),
            Coco17.RIGHT_WRIST to Keypoint2D(0.65f, 0.40f, 0.1f), // gated
            Coco17.RIGHT_HIP to Keypoint2D(0.50f, 0.55f, 1f),
            Coco17.LEFT_HIP to Keypoint2D(0.48f, 0.55f, 1f),
            Coco17.RIGHT_KNEE to Keypoint2D(0.50f, 0.72f, 1f),
            Coco17.RIGHT_ANKLE to Keypoint2D(0.52f, 0.90f, 1f)
        )
        val m = DrillMetrics.extractAtFrame(f, Handedness.RIGHT, 1f)
        assertFalse(DrillMetrics.METRIC_ELBOW_ANGLE in m, "elbow needs the wrist")
        assertTrue(DrillMetrics.METRIC_KNEE_BEND in m, "knee unaffected by wrist score")
    }

    @Test
    fun insaneValueIsDropped() {
        // Fully straight arm = 180° elbow — outside the 20–170° sanity band → dropped
        val f = frame(
            Coco17.RIGHT_SHOULDER to Keypoint2D(0.50f, 0.20f, 1f),
            Coco17.RIGHT_ELBOW to Keypoint2D(0.50f, 0.40f, 1f),
            Coco17.RIGHT_WRIST to Keypoint2D(0.50f, 0.60f, 1f)
        )
        val m = DrillMetrics.extractAtFrame(f, Handedness.RIGHT, 1f)
        assertFalse(DrillMetrics.METRIC_ELBOW_ANGLE in m)
    }

    @Test
    fun emptyFrameYieldsNoMetrics() {
        assertTrue(DrillMetrics.extractAtFrame(PoseFrame2D(0, 0L, emptyList()), Handedness.RIGHT, 1f).isEmpty())
    }

    @Test
    fun sanityBoundsSpotChecks() {
        assertTrue(SanityBounds.isSane(DrillMetrics.METRIC_ELBOW_ANGLE, 90.0))
        assertFalse(SanityBounds.isSane(DrillMetrics.METRIC_ELBOW_ANGLE, 19.0))
        assertFalse(SanityBounds.isSane(DrillMetrics.METRIC_ELBOW_ANGLE, 171.0))
        assertTrue(SanityBounds.isSane("unknown_metric", 12345.0), "unknown metrics pass through")
    }

    /** 90°-elbow frame used by the peak-window tests. */
    private fun goodArmFrame(idx: Int, wrist: Keypoint2D = Keypoint2D(0.70f, 0.42f, 1f)): PoseFrame2D {
        val f = frame(
            Coco17.RIGHT_SHOULDER to Keypoint2D(0.50f, 0.22f, 1f),
            Coco17.RIGHT_ELBOW to Keypoint2D(0.50f, 0.42f, 1f),
            Coco17.RIGHT_WRIST to wrist
        )
        return f.copy(frameIndex = idx, timestampMs = idx * 33L)
    }

    @Test
    fun peakMetricsAreMedianSmoothedOverTheWindow() {
        // L-05: middle frame (the peak) has a jittered wrist that alone reads ~135°;
        // the ±70 ms window at 33 ms holds 5 frames and the median must ignore it.
        val frames = listOf(
            goodArmFrame(0), goodArmFrame(1),
            goodArmFrame(2, wrist = Keypoint2D(0.64f, 0.56f, 1f)), // jitter → ≈135°
            goodArmFrame(3), goodArmFrame(4)
        )
        val m = DrillMetrics.extractAtPeak(frames, peakFrame = 2, Handedness.RIGHT, 1f, intervalMs = 33L)
        assertEquals(90.0, m[DrillMetrics.METRIC_ELBOW_ANGLE]!!, 1.0)
    }

    @Test
    fun peakWindowDegradesToSingleFrameAtCoarseIntervals() {
        // radius = 70 ms / 100 ms = 0 frames → identical to extractAtFrame
        val frames = listOf(goodArmFrame(0), goodArmFrame(1), goodArmFrame(2))
        val atPeak = DrillMetrics.extractAtPeak(frames, peakFrame = 1, Handedness.RIGHT, 1f, intervalMs = 100L)
        val atFrame = DrillMetrics.extractAtFrame(frames[1], Handedness.RIGHT, 1f)
        assertEquals(atFrame.keys, atPeak.keys)
        atFrame.forEach { (k, v) -> assertEquals(v, atPeak[k]!!, 1e-9) }
    }
}

File: poses_viewer/src/drill2d/__tests__/drillMetrics.test.ts
/**
 * drillMetrics.ts — mirrors Kotlin DrillMetricsTest.
 *
 * Key parity notes vs the original task spec:
 *  - A fully straight arm (180°) is OUTSIDE the [20, 170] sanity band and correctly
 *    dropped. Tests use a ~90° elbow (shoulder above elbow, wrist to the right) so
 *    the angle lands inside the sane range.
 *  - The "drops when out of sanity bounds" case uses a wrist position that yields
 *    an insane angle (≈180° straight arm), not a low-score gate — matching Kotlin's
 *    `insaneValueIsDropped` test.
 */
import { describe, expect, it } from 'vitest'
import { extractAtFrame, extractAtPeak, METRIC, ALL_KEYS } from '../drillMetrics'
import { Coco17, Keypoint2D, PoseFrame2D } from '../types'
import { METRIC_KEYS } from '../referenceStandard'

const K = (x: number, y: number, score = 1): Keypoint2D => ({ x, y, score })

/**
 * Frame with a ~90° right elbow: shoulder directly above elbow, wrist to the right.
 * All other keypoints default to (0.5, 0.5, score=1) so they don't accidentally
 * gate torso / knee metrics.
 */
function goodArmFrame(idx: number, wrist: Keypoint2D = K(0.70, 0.42)): PoseFrame2D {
  const kp: Keypoint2D[] = Array.from({ length: 17 }, () => K(0.5, 0.5))
  kp[Coco17.RIGHT_SHOULDER] = K(0.50, 0.22)
  kp[Coco17.RIGHT_ELBOW]    = K(0.50, 0.42)
  kp[Coco17.RIGHT_WRIST]    = wrist
  return { frameIndex: idx, timestampMs: idx * 33, keypoints: kp }
}

/** Fully straight arm (180°) — outside the [20,170] sanity band → dropped. */
function straightArmFrame(idx: number): PoseFrame2D {
  const kp: Keypoint2D[] = Array.from({ length: 17 }, () => K(0.5, 0.5))
  kp[Coco17.RIGHT_SHOULDER] = K(0.50, 0.20)
  kp[Coco17.RIGHT_ELBOW]    = K(0.50, 0.40)
  kp[Coco17.RIGHT_WRIST]    = K(0.50, 0.60)
  return { frameIndex: idx, timestampMs: idx * 33, keypoints: kp }
}

describe('extractAtFrame', () => {
  it('extracts elbow angle when the arm is bent within sanity bounds (~90°)', () => {
    const m = extractAtFrame(goodArmFrame(0), 'right', 1)
    expect(m[METRIC.ELBOW_ANGLE]).toBeDefined()
    expect(m[METRIC.ELBOW_ANGLE]).toBeGreaterThan(20)
    expect(m[METRIC.ELBOW_ANGLE]).toBeLessThan(170)
  })

  it('drops elbow angle when the arm is fully straight (180° > sanity upper bound 170°)', () => {
    const m = extractAtFrame(straightArmFrame(0), 'right', 1)
    expect(m[METRIC.ELBOW_ANGLE]).toBeUndefined()
  })

  it('drops elbow angle when the wrist keypoint is below minScore', () => {
    const kp: Keypoint2D[] = Array.from({ length: 17 }, () => K(0.5, 0.5))
    kp[Coco17.RIGHT_SHOULDER] = K(0.50, 0.22)
    kp[Coco17.RIGHT_ELBOW]    = K(0.50, 0.42)
    kp[Coco17.RIGHT_WRIST]    = K(0.70, 0.42, 0.1) // score below DEFAULT_MIN_SCORE
    const frame: PoseFrame2D = { frameIndex: 0, timestampMs: 0, keypoints: kp }
    expect(extractAtFrame(frame, 'right', 1)[METRIC.ELBOW_ANGLE]).toBeUndefined()
  })

  it('returns empty map for a frame with no keypoints', () => {
    const frame: PoseFrame2D = { frameIndex: 0, timestampMs: 0, keypoints: [] }
    expect(Object.keys(extractAtFrame(frame, 'right', 1))).toHaveLength(0)
  })
})

describe('extractAtPeak', () => {
  it('returns median ~90° over a ±70ms window of identical good frames', () => {
    const frames = [0, 1, 2, 3, 4].map(i => goodArmFrame(i))
    const m = extractAtPeak(frames, 2, 'right', 1, 33, undefined, 70)
    expect(m[METRIC.ELBOW_ANGLE]).toBeDefined()
    expect(m[METRIC.ELBOW_ANGLE]).toBeGreaterThan(20)
    expect(m[METRIC.ELBOW_ANGLE]).toBeLessThan(170)
  })

  it('ignores a jitter frame at the peak when neighbours are consistent', () => {
    // frame 2 (peak) has a wrist that reads ~135°; the other 4 frames are ~90°
    // median of [90, 90, 135, 90, 90] = 90 → jitter suppressed
    const jitteredWrist = K(0.64, 0.56)
    const frames = [
      goodArmFrame(0), goodArmFrame(1),
      goodArmFrame(2, jitteredWrist),
      goodArmFrame(3), goodArmFrame(4),
    ]
    const m = extractAtPeak(frames, 2, 'right', 1, 33, undefined, 70)
    expect(m[METRIC.ELBOW_ANGLE]).toBeCloseTo(
      extractAtFrame(goodArmFrame(0), 'right', 1)[METRIC.ELBOW_ANGLE]!,
      0,
    )
  })

  it('degrades to a single frame when intervalMs > radiusMs (radius = 0)', () => {
    const frames = [goodArmFrame(0), goodArmFrame(1), goodArmFrame(2)]
    const atPeak  = extractAtPeak(frames, 1, 'right', 1, 100, undefined, 70)
    const atFrame = extractAtFrame(frames[1], 'right', 1)
    for (const key of Object.keys(atFrame)) {
      expect(atPeak[key]).toBeCloseTo(atFrame[key], 9)
    }
  })

  it('throws on a non-positive intervalMs', () => {
    const frames = [goodArmFrame(0)]
    expect(() => extractAtPeak(frames, 0, 'right', 1, 0)).toThrow()
  })

  it('throws when peakFrame is out of bounds', () => {
    const frames = [goodArmFrame(0), goodArmFrame(1)]
    expect(() => extractAtPeak(frames, 2, 'right', 1, 33)).toThrow()
    expect(() => extractAtPeak([], 0, 'right', 1, 33)).toThrow()
  })

  it('ALL_KEYS lists the six metric keys', () => {
    expect(ALL_KEYS).toHaveLength(6)
    expect(ALL_KEYS).toContain(METRIC.ELBOW_ANGLE)
    expect(ALL_KEYS).toContain(METRIC.SHOULDER_ANGLE)
    expect(ALL_KEYS).toContain(METRIC.KNEE_BEND)
    expect(ALL_KEYS).toContain(METRIC.TORSO_LEAN)
    expect(ALL_KEYS).toContain(METRIC.SHOULDER_TILT)
    expect(ALL_KEYS).toContain(METRIC.HIP_FLEXION)
  })
})

describe('metric key consistency', () => {
  it('ALL_KEYS matches referenceStandard.METRIC_KEYS (no drift between the two definitions)', () => {
    expect([...ALL_KEYS].sort()).toEqual([...METRIC_KEYS].sort())
  })
})
